### ISSUE TITLE
End of sequence after a few batch with dataset.shard

### ISSUE THREAD

**Issue by malomarrec** [tensorflow/tensorflow#18040]
When running distributed training, my session gets terminated after a few batches when using `dataset.shard`. The issue disappears when I run distributed training with an independent dataset object on each worker (in which case the same data gets shuffled and read on each worker).

TensorFlow version: TF 1.6, GPU

Dataset construction:
```
def construct_dataset(filenames, labels, batch_size, num_workers,worker_index):
    dataset = tf.data.TextLineDataset(filenames)
    dataset = tf.data.Dataset.from_tensor_slices((filenames,labels))
    dataset = dataset.shard(num_workers, worker_index) #works fine when commenting this out
    dataset = dataset.shuffle(buffer_size=10000)  # Equivalent to min_after_dequeue=10000.
    dataset = dataset.map(_parse_function)
    dataset = dataset.batch(batch_size)
    return dataset
```

main
```
with tf.device(device):
        filelist, labels = get_filelist(FLAGS.data_dir)
        dataset = construct_dataset(...)
        iterator = dataset.make_one_shot_iterator()
        batch = iterator.get_next()
        img_batch, filepath_batch, label_batch = batch
        ...

       hooks=[tf.train.StopAtStepHook(last_step=1000000)] 

    with tf.train.MonitoredTrainingSession(master=target,
        is_chief=(FLAGS.task_index == 0),checkpoint_dir=logs,hooks = hooks) as sess:
        try:
            while not sess.should_stop():
                sess.run(train_op)
        except Exception as e:
            print(e)


Stacktrace - 
```End of sequence
	 [[Node: IteratorGetNext = IteratorGetNext[output_shapes=[[?,300,300,?], [?], [?]], output_types=[DT_UINT8, DT_STRING, DT_INT32], _device="/job:ps/replica:0/task:0/device:CPU:0"](OneShotIterator)]]
	 [[Node: Training_Loss_S17 = _HostRecv[client_terminated=false, recv_device="/job:worker/replica:0/task:0/device:CPU:0", send_device="/job:ps/replica:0/task:0/device:CPU:0", send_device_incarnation=-379211586651304706, tensor_name="edge_133_Training_Loss", tensor_type=DT_STRING, _device="/job:worker/replica:0/task:0/device:CPU:0"]()]]
```

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by malomarrec** [comment]
Done.

**Comment by asimshankar** [comment]
I apologize, but I don't quite understand the issue here. Could you provide the full stack trace?

Also, a side observation, the first line in `construct_dataset` seems unnecessary, since the very next line ignores it and creates a new `Dataset` using `from_tensor_slices`?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
It has been 29 days with no activity and the `awaiting response` label was assigned. Is this still an issue?
